src/som/interpreter/nodes/ResolvingImplicitReceiverSend.java

```diff
@@ -73,38 +73,49 @@ public Object doPreEvaluated(final VirtualFrame frame, final Object[] args) {
     Lock lock = getLock();
     try {
       lock.lock();
-      newNode = specialize(args);
+      newNode = specialize(frame, args);
     } finally {
       lock.unlock();
     }
     return newNode.doPreEvaluated(frame, args);
   }
 
-  private PreevaluatedExpression specialize(final Object[] args) {
+  private PreevaluatedExpression specialize(final VirtualFrame frame, final Object[] args) {
     if (replacedBy != null) {
       // already has been specialized
       if (newReceiverNodeForOuterSend != null) {
         // need to recalculate the real receiver for outer sends
-        args[0] = newReceiverNodeForOuterSend.executeEvaluated(args[0]);
+        args[0] = newReceiverNodeForOuterSend.executeGeneric(frame);
```

Comment:
@richard-roberts While I agree that we need this semantics, the change itself is highly problematic, because it relies on complex assumptions to be true.
Generally, we can not reevaluate any nodes arbitrarily, because they might have side effects that are visible on the language level.

Here the key assumption is that the evaluation of the outer chain won't cause side effects.
Specifically, the evaluation of the original `args[0]` is not allowed to have side effects, so that a reevaluation is unproblematic. I have been thinking of ensuring this with assertions, which seems hard. So, I'll put in explicit code that guarantees this.